### ISSUE TITLE
ActorContext is null after supervisor restart

### ISSUE THREAD

**Issue by SmedbergM** [akka/akka#29995]
After a supervisor restart, an actor's `context` is null. This violates the thread-safety guarantees of `context`, which include a guarantee that `context.parent` will be provide thread-safe access. See https://github.com/SmedbergM/akka-parent-npe for a working example.

# Steps to reproduce

Start an actor which spawns child actors. Schedule events in the future which close over the child actor's `context` and will send a message to `context.parent`. Then trigger a failure in the root actor, causing a restart.

## Expected behavior

Any messages sent to the parent are sent to the restarted parent. Any messages sent to an intermediate actor are sent to the dead-letter bin.

## Actual behavior
`context` is null, resulting in both messages throwing a `NullPointerException`.

**Comment by johanandren** [comment]
The guarantee mentioned in the API-doc is that if you have a `ActorContext`/`ActorContext[T]` it is safe to access `parent` from other threads, not that the field `context` in the actor will be available for all foreseeable future regardless of if the actor lifecycle. The API doc of `Actor#context` actually even stresses this:

>   * WARNING: Only valid within the Actor itself, so do not close over it and
>   * publish it to other threads!

If you want to do that you will have to close over the value of `context` (or perhaps better, `context.parent`) at the time when the actor is still running and pass that to the logic you will execute in a different thread. In general we recommend to put logic that can potentially close over `context` and other actor state unsafely in static/companion object methods to make their dependencies explicit and avoiding bugs.

**Comment by patriknw** [comment]
@SmedbergM I agree with Johan's recommendation about being explicit and minimize what is closed over.

A good convention is to place all such callbacks (futures, streams, scheduler callbacks) outside of the actor instance. A method in the companion object for example.

That said, I think we have fixed this particular problem in later versions of Akka. Try with latest Akka 2.6.12. It was fixed in https://github.com/akka/akka/issues/25040
